### ISSUE TITLE
The deferred returned by twisted.names.client.Resolver.queryTCP never fires if the TCP connection attempt fails

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6658]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#6658|
|Type|defect|
|Created|2013-08-02 00:34:07Z|
|Branch|https://github.com/twisted/twisted/tree/querytcp-connect-failure-6658|

If the TCP connection is refused or dropped, the original deferred returned by queryTCP should errback with the reason for the connection failure.

Here's a failing test:

```
Index: twisted/names/test/test_client.py
===================================================================
diff --git a/branches/opt-record-5668-4/twisted/names/test/test_client.py b/branches/opt-record-5668-4/twisted/names/test/test_client.py
--- a/branches/opt-record-5668-4/twisted/names/test/test_client.py	(revision 39321)
+++ b/branches/opt-record-5668-4/twisted/names/test/test_client.py	(working copy)
@@ -24,6 +24,8 @@
 from twisted.names.test.test_hosts import GoodTempPathMixin
 from twisted.names.test.test_rootresolve import MemoryReactor
 
+from twisted.test import proto_helpers
+
 from twisted.trial import unittest
 
 if platform.isWindows():
@@ -667,7 +669,29 @@
         self.assertNotIn(protocol, resolver.connections)
 
 
+    def test_queryTCPDeferredErrbackOnConnectionFailure(self):
+        """
+        The deferred returned by L{client.Resolver.queryTCP} will
+        errback when the TCP connection attempt fails. The reason for
+        the connection failure is passed to error handlers on all the
+        original deferreds.
+        """
+        reactor = proto_helpers.MemoryReactor()
+        resolver = client.Resolver(
+            servers=[('192.0.2.100', 53)],
+            reactor=reactor)
 
+        d1 = resolver.queryTCP(dns.Query('example.com'))
+        d2 = resolver.queryTCP(dns.Query('example.net'))
+        host, port, factory, timeout, bindAddress = reactor.tcpClients[0]
+        class SentinelException(Exception):
+            pass
+        factory.clientConnectionFailed(reactor.connectors[0], SentinelException())
+        self.failureResultOf(d1, SentinelException)
+        self.failureResultOf(d2, SentinelException)
+
+
+
 class ClientTestCase(unittest.TestCase):
 
     def setUp(self):
```

...and a proposed fix...

```
Index: twisted/names/client.py
===================================================================
diff --git a/branches/opt-record-5668-4/twisted/names/client.py b/branches/opt-record-5668-4/twisted/names/client.py
--- a/branches/opt-record-5668-4/twisted/names/client.py	(revision 39321)
+++ b/branches/opt-record-5668-4/twisted/names/client.py	(working copy)
@@ -508,6 +508,11 @@
         pass
 
 
+    def clientConnectionFailed(self, connector, reason):
+        for d, q, t in self.controller.pending:
+            d.errback(reason)
+
+
     def buildProtocol(self, addr):
         p = dns.DNSProtocol(self.controller)
         p.factory = self

```

<details><summary>Searchable metadata</summary>

```
trac-id__6658 6658
type__defect defect
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_querytcp_connect_failure_6658 branches/querytcp-connect-failure-6658
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1375403647000000 1375403647000000
changetime__1376145574000000 1376145574000000
version__None None
owner__rwall rwall

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_1"></a>|@wallrj commented|
|-|-|

(In [39417]) Branching to 'querytcp-connect-failure-6658'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_2"></a>|@wallrj commented|
|-|-|

Ready for review in log:branches/querytcp-connect-failure-6658

If this gets merged I'll be able to fix [#6659](https://github.com/twisted/twisted/issues/6659) next.

Build Results:
 * http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/querytcp-connect-failure-6658

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @wallrj|
|-|-|

Thanks for working on this.

  1. in the `clientConnectionFailed` docstring, I'd either say that the method _fails_ queries or _errbacks_ deferreds, not errbacks queries.
  1. The TCP _connection_ is probably what fails, rather than the TCP _connect_ (and same for the news fragment).
  1. I don't think the documentation for `reason` needs to be so specific about how the implementation works.  It's too specific, in fact - if you're using iocpreactor  then `twisted.internet.tcp.Client` isn't what drives this process.
  1. Similar comment for the `connector` parameter.  There's actually an interface for this one, `IConnector`.  But for both these parameters, it might be better to just refer readers to `twisted.internet.protocol.ClientFactory.clientConnectionLost`.  The main thing that's interesting about this method is that it does pass the reason on to whatever deferreds it fails.
  1. What happens if one of the errbacks on a queryTCP deferred calls queryTCP again?  I think there's a reentrancy bug in the loop in `clientConnectionLost`.
  1. In `test_queryTCPDeferredErrbackOnConnectionFailure`, I think you're issuing two queries to exercise the loop logic.  Can you split this into two tests, one for the simpler case of a single query and one that asserts something about what happens to the second query?

Thanks!  Please merge after addressing those points.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj set status to `assigned`|
|-|-|

Replying to [exarkun](#note_3):
> Thanks for working on this.
>   1. in the `clientConnectionFailed` docstring, I'd either say that
>   the method _fails_ queries or _errbacks_ deferreds, not
>   errbacks queries.
>   1. The TCP _connection_ is probably what fails, rather than the
>   TCP _connect_ (and same for the news fragment).

Done. I said "connection attempt" rather than "connection".  I
interpret "connection" as an established connection.

>   1. I don't think the documentation for `reason` needs to be so
>   specific about how the implementation works.  It's too specific,
>   in fact - if you're using iocpreactor  then
>   `twisted.internet.tcp.Client` isn't what drives this process.
>   1. Similar comment for the `connector` parameter.  There's
>   actually an interface for this one, `IConnector`.  But for both
>   these parameters, it might be better to just refer readers to
>   `twisted.internet.protocol.ClientFactory.clientConnectionLost`.
>   The main thing that's interesting about this method is that it
>   does pass the reason on to whatever deferreds it fails.

I improved the documentation.

I linked to
twisted.internet.protocol.ClientFactory.clientConnectionLost too, but
that doesn't actually have any parameter documentation.

>   1. What happens if one of the errbacks on a queryTCP deferred
>   calls queryTCP again?  I think there's a reentrancy bug in the
>   loop in `clientConnectionLost`.

Fixed....I think. r39437.

That's an example of a test where I think I do need multiple assertions in
the test.

Also it seems slightly ugly to be mangling the internals of
client.Resolver from client.DNSClientFactory. Maybe it could dispatch
to a new method eg client.Resolver.connectionFailed

>   1. In `test_queryTCPDeferredErrbackOnConnectionFailure`, I think
>   you're issuing two queries to exercise the loop logic.  Can you
>   split this into two tests, one for the simpler case of a single
>   query and one that asserts something about what happens to the
>   second query?

Done.

>
> Thanks!  Please merge after addressing those points.

New build results:
 * http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/querytcp-connect-failure-6658

I'll merge tomorrow morning if build succeeds and unless you have any
further comments about the changes I made.

Thanks for the review.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj set status to `closed`|
|-|-|

(In [39546]) Merge querytcp-connect-failure-6658

Author: rwall
Reviewer: exarkun
Fixes: [#6658](https://github.com/twisted/twisted/issues/6658)

The Deferred returned by twisted.names.client.Resolver.queryTCP now
fires with an error if the TCP connection attempt fails.
